apps/aecore/src/aec_trees.erl

```diff
@@ -204,38 +208,111 @@ contracts(Trees) ->
 set_contracts(Trees, Contracts) ->
     Trees#trees{contracts = Contracts}.
 
+%%%=============================================================================
+%%% Serialization for db storage
+%%%=============================================================================
+
+-define(AEC_TREES_VERSION, 0).
+
+-spec deserialize_from_db(binary()) -> trees().
+deserialize_from_db(Bin) when is_binary(Bin) ->
+    [ {contracts_hash, Contracts}
+    , {calls_hash, Calls}
+    , {channels_hash, Channels}
+    , {ns_hash, NS}
+    , {ns_cache_hash, NSCache}
+    , {oracles_hash, Oracles}
+    , {oracles_cache_hash, OraclesCache}
+    , {accounts_hash, Accounts}
+    ] = lists:map(fun db_deserialize_hash/1,
+                  aec_object_serialization:deserialize(
+                    trees_db,
+                    ?AEC_TREES_VERSION,
+                    db_serialization_template(?AEC_TREES_VERSION),
+                    Bin
+                   )),
+    #trees{ contracts = aect_state_tree:new_with_backend(Contracts)
+          , calls     = aect_call_state_tree:new_with_backend(Calls)
+          , channels  = aesc_state_tree:new_with_backend(Channels)
+          , ns        = aens_state_tree:new_with_backend(NS, NSCache)
+          , oracles   = aeo_state_tree:new_with_backend(Oracles, OraclesCache)
+          , accounts  = aec_accounts_trees:new_with_backend(Accounts)
+          }.
+
+-spec serialize_for_db(trees()) -> binary().
+serialize_for_db(#trees{} = Trees) ->
+    aec_object_serialization:serialize(
+      trees_db,
+      ?AEC_TREES_VERSION,
+      db_serialization_template(?AEC_TREES_VERSION),
+      lists:map(fun db_serialize_hash/1,
+                [ {contracts_hash,     contracts_hash(Trees)}
+                , {calls_hash,         calls_hash(Trees)}
+                , {channels_hash,      channels_hash(Trees)}
+                , {ns_hash,            ns_hash(Trees)}
+                , {ns_cache_hash,      ns_cache_hash(Trees)}
+                , {oracles_hash,       oracles_hash(Trees)}
+                , {oracles_cache_hash, oracles_cache_hash(Trees)}
+                , {accounts_hash,      accounts_hash(Trees)}
+                ])
+     ).
+
+db_serialization_template(?AEC_TREES_VERSION) ->
+    [ {contracts_hash,     [binary]}
+    , {calls_hash,         [binary]}
+    , {channels_hash,      [binary]}
+    , {ns_hash,            [binary]}
+    , {ns_cache_hash,      [binary]}
+    , {oracles_hash,       [binary]}
+    , {oracles_cache_hash, [binary]}
+    , {accounts_hash,      [binary]}
```

Comment:
I see (from 3 lines below, and from removal of hash padding) that it is list of binary as it could be one hash or none - in case tree empty. Ok.